transformer_engine/common/fused_attn/thd_utils.h

```diff
@@ -102,58 +102,187 @@ __global__ void thd_lse_kernel(lse_dtype *lse, float *half_lse, int *cu_seqlens,
+  int8_t store_format[4];
+  int *cu_seqlens_s;
+  // size of tensor, b s h t
+  int size[4];
+  __forceinline__ __device__ TensorFormat(int size_kernel[4], int *cu_seqlens = nullptr) {
+    for (int i = 0; i < 4; i++) {
+      size[i] = size_kernel[i];
+    }
+
+    if constexpr (format == QKVFormat::TH) {
+      cu_seqlens_s = cu_seqlens;
+      store_format[0] = 3;
+      store_format[1] = 2;
+      store_format[3] = 1;
+    } else if constexpr (format == QKVFormat::HT) {
+      cu_seqlens_s = cu_seqlens;
+      store_format[0] = 2;
+      store_format[1] = 3;
+      store_format[3] = 1;
+    } else if constexpr (format == QKVFormat::SBH) {
+      store_format[0] = 1;
+      store_format[1] = 0;
+      store_format[2] = 2;
+      store_format[3] = 0;
+    } else if constexpr (format == QKVFormat::HBS) {
+      store_format[0] = 2;
+      store_format[1] = 0;
+      store_format[2] = 1;
+      store_format[3] = 0;
+    } else if constexpr (format == QKVFormat::BSH) {
+      store_format[0] = 0;
+      store_format[1] = 1;
+      store_format[2] = 2;
+      store_format[3] = 0;
+    } else if constexpr (format == QKVFormat::BHS) {
+      store_format[0] = 0;
+      store_format[1] = 2;
+      store_format[2] = 1;
+      store_format[3] = 0;
+    }
+  }
+
+  // calculate address according to index
+  __forceinline__ __device__ int compute_address(int id[4]) {
+    int address;
+    if (store_format[3] == 1) {
+      address = id[store_format[0]] * size[store_format[1]] + id[store_format[1]];
+    } else {
+      address = id[store_format[0]] * size[store_format[1]] + id[store_format[1]];
+      address = address * size[store_format[2]] + id[store_format[2]];
+    }
+    return address;
+  }
+
+  // compute half right index
+  __forceinline__ __device__ void compute_half_right(int id[4]) {
+    if constexpr (format == QKVFormat::TH) {
+      id[1] -= (cu_seqlens_s[id[0] + 1] - cu_seqlens_s[id[0]]) / 2;
+      id[3] -= cu_seqlens_s[id[0] + 1] / 2;
+    } else if constexpr (format == QKVFormat::BSH || format == QKVFormat::SBH) {
+      id[1] -= size[1] / 2;
+    }
+  }
+};
+
+template <typename dtype, int tile_size, bool causal, QKVFormat out_format, QKVFormat lse_format,
+          int max_tensors>
+__global__ void fused_out_correction_kernel(dtype *out, TensorList<max_tensors> tensors, float *lse,
+                                            int *cu_seqlens, int batch, int num_heads,
+                                            int dim_per_head, int lse_seqlen, int cp_size, int rank,
+                                            int start) {
   extern __shared__ int cu_seqlens_s[];
-  for (int i = threadIdx.x; i <= batch; i += blockDim.x) {
-    cu_seqlens_s[i] = cu_seqlens[i] / (only_second_half + 1);
+  int full_num;
+  int num_total_tokens;
+
+  if constexpr (out_format == QKVFormat::TH) {
+    for (int i = threadIdx.x; i <= batch; i += blockDim.x) {
+      cu_seqlens_s[i] = cu_seqlens[i];
```

Comment:
why is this needed?

Reply:
we need this to initialize TensorFormat when THD format is applied